presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java

```diff
@@ -631,13 +631,13 @@ private boolean isRecoverable(List<ExecutionFailureInfo> failures)
 
     private synchronized void updateFinalTaskInfo(TaskInfo finalTaskInfo)
     {
-        tasksWithFinalInfo.add(finalTaskInfo.getTaskId());
+        tasksWaitingForFinalInfo.remove(finalTaskInfo.getTaskId());
         checkAllTaskFinal();
     }
 
     private synchronized void checkAllTaskFinal()
     {
-        if (stateMachine.getState().isDone() && tasksWithFinalInfo.containsAll(allTasks)) {
+        if (stateMachine.getState().isDone() && tasksWaitingForFinalInfo.isEmpty()) {
```

Comment:
I am finding this tasksWaitingForFinalInfo hard to grok. Would it be easier to reword it as nonFinalTasks ?

Reply:
+1, or just `runningTasks`.